### PR TITLE
fix: Broken reschedule link in email created by workflow for team event

### LINKED ISSUE #13869: [CAL-3216] Broken reschedule link in email created by workflow for team event
### Issue Summary

The reschedule link in email created by the workflow for a team event doesn't work because its slug doesn't contain `/team`.

### Steps to Reproduce

1. Create an event type _for a team_.
2. Assign a workflow to the event type that sends an email containing the `{RESCHEDULE_URL}` variable.
3. Book the team event (`cal.com/team/team-name/event-name`).
5. Receive an email containing a reschedule link for the event via the workflow.
6. Click on the reschedule link in the email.

### Actual Results

The reschedule link in the email looks like `app.cal.com/team-name/event-name?rescheduleUid=...`. Clicking on it results in a 404 page because `/team` is missing in the slug.

### Expected Results

The reschedule link in the email should look like `app.cal.com/team/team-name/event-name?rescheduleUid=...` and open the reschedule page of the event.

<sub>[CAL-3216](https://linear.app/calcom/issue/CAL-3216/broken-reschedule-link-in-email-created-by-workflow-for-team-event)</sub>

### LINKED ISSUE #13885: [CAL-3219] [CAL-3218] {RESCHEDULE_URL} is returning "undefined" in the url and not working
**Steps to recreate**

* Create a workflow with a custom message 
* Add a {RESCHEDULE_URL}
* View the email that is sent: It returns a link that does not worl

CleanShot 2024-02-27 at 11 34 39@2x

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3218](https://linear.app/calcom/issue/CAL-3218/reschedule-url-is-returning-undefined-in-the-url-and-not-working)</sub>

<sub>[CAL-3219](https://linear.app/calcom/issue/CAL-3219/[cal-3218]-reschedule-url-is-returning-undefined-in-the-url-and-not)</sub>

### PR DESCRIPTION
**Pull request by Pritam-sd-dev**
## What does this PR do?

Fixes #13869 
Fixed #13885

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?
add a workflow to event with variable rescheduleLink and create a event check link in the email.

## Checklist
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by CarinaWolli — DISMISSED** [review]
We also need to fix the reschedule URL for orgs, it should show the correct org url like `acme.cal.com/username/event-type`. Could you take a look at that as well?

**Comment by Pritam-sd-dev** [comment]
> We also need to fix the reschedule URL for orgs, it should show the correct org url like `acme.cal.com/username/event-type`. Could you take a look at that as well?

yes, sure i will lookup

**Comment by Pritam-sd-dev** [comment]
@CarinaWolli 
as of now webappurl is hardcoded, according to me its wrong because for orgs
link should me like org-slug.cal.com/username/event-type correct ?

**Comment by Pritam-sd-dev** [comment]
@CarinaWolli 
username is inserted in link correctly in my case.
Screenshot from 2024-02-28 12-16-34
not able to reproduce bug.
can you please provide more context for it ?

**Comment by CarinaWolli** [comment]
> as of now webappurl is hardcoded, according to me its wrong because for orgs link should be like org-slug.cal.com/username/event-type correct ?

Exactly, so on local it's right now: 
 http://app.cal.local:3000/owner1-acme/owner1-acme-event?rescheduleUid=
but should be: 
 http://acme.cal.local:3000/owner1-acme/owner1-acme-event?rescheduleUid=

> username is inserted in link correctly in my case.

Not sure what you mean by that, is that org related as well?

**Comment by Pritam-sd-dev** [comment]
@CarinaWolli 
i have implemented logic for rescheduleLink fix at one place, please review that code so that i can implement to other required places.

File: packages/features/ee/workflows/api/scheduleSMSReminders.ts [hunk 1]
```diff
@@ -99,6 +101,37 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
           booking: reminder.booking,
         });
 
+        // checks whether booking belongs to a organization and updates rescheduleLink.
+        let rescheduleLink = WEBAPP_URL;
```
Comment (CarinaWolli) [inline packages/features/ee/workflows/api/scheduleSMSReminders.ts:105]: I think we can do this a lot easier and use the same logic as in handleNewBooking.ts to get the bookerUrl, see here: https://github.com/calcom/cal.com/blob/main/packages/features/bookings/lib/handleNewBooking.ts#L1452

**Comment by CarinaWolli** [comment]
@Pritam-sd-dev were you able to reproduce the issue with undefined in the reschedule link? https://github.com/calcom/cal.com/issues/13885

I think the case shown in the screenshot was a round robin team event

**Comment by CarinaWolli** [comment]
Just pushed some changes to support org URLs, there are still some type errors I need to fix

**Review by CarinaWolli — COMMENTED** [review]
I found some unnecessary complex type structures for workflows when working on that part of the code, I'd like to revisit that in a separate PR

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 2]
```diff
@@ -1480,7 +1481,7 @@ async function handler(
       id: organizerUser.id,
       name: organizerUser.name || "Nameless",
       email: organizerEmail,
-      username: organizerUser.username || undefined,
+      username: organizerOrganizationProfile?.username || organizerUser.username || undefined,
```
[the remark this answers is not in the document]
Comment (Udit-takkar) [reply packages/features/bookings/lib/handleNewBooking.ts:1484]: Yeah this is confusing to me.

File: packages/features/bookings/lib/handleCancelBooking.ts [hunk 3]
```diff
@@ -263,7 +263,12 @@ async function handler(req: CustomRequest) {
       : [],
     cancellationReason: cancellationReason,
     ...(teamMembers && {
-      team: { name: bookingToDelete?.eventType?.team?.name || "Nameless", members: teamMembers, id: teamId! },
+      team: {
+        name: bookingToDelete?.eventType?.team?.name || "Nameless",
+        members: teamMembers,
+        id: teamId!,
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleCancelBooking.ts:269]: NIT: add check for teamId 

...(teamMembers && teamId

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 4]
```diff
@@ -173,8 +186,10 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
             additionalNotes: reminder.booking.description,
             responses: responses,
             meetingUrl: bookingMetadataSchema.parse(reminder.booking.metadata || {})?.videoCallUrl,
-            cancelLink: `/booking/${reminder.booking.uid}?cancel=true`,
-            rescheduleLink: `/${reminder.booking.user?.username}/${reminder.booking.eventType?.slug}?rescheduleUid=${reminder.booking.uid}`,
+            cancelLink: `${bookerUrl}/booking/${reminder.booking.uid}?cancel=true`,
+            rescheduleLink: `${bookerUrl}/${
+              organizerOrganizationProfile?.username || reminder.booking.user?.username
+            }/${reminder.booking.eventType?.slug}?rescheduleUid=${reminder.booking.uid}`,
```
Comment (Udit-takkar) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:192]: This is the reschedule url used in Confirmation email packages/lib/CalEventParser.ts

` `${calEvent.bookerUrl ?? WEBAPP_URL}/reschedule/${seatUid ? seatUid : Uid}`;`

CancelLink is also different

File: packages/features/ee/workflows/api/scheduleSMSReminders.ts [hunk 5]
```diff
@@ -91,14 +85,26 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
...
+            userId: reminder.booking.user?.id,
+          },
+        });
+
+        const organizerOrganizationId = organizerOrganizationProfile?.organizationId;
+
+        const bookerUrl = reminder.booking.eventType?.team
+          ? await getBookerBaseUrl(reminder.booking.eventType?.team.parentId ?? null)
+          : await getBookerBaseUrl(organizerOrganizationId ?? null);
+
```
Comment (Udit-takkar) [inline packages/features/ee/workflows/api/scheduleSMSReminders.ts:107]: NIT

```suggestion
        const bookerUrl =  await getBookerBaseUrl(reminder.booking.eventType?.team?.parentId ?? organizerOrganizationId ?? null)
    

```

File: packages/lib/payment/getBooking.ts [hunk 6]
```diff
@@ -82,11 +121,24 @@ export async function getBooking(bookingId: number) {
...
+    where: {
+      userId: booking.userId ?? undefined,
+    },
+  });
+
+  const organizerOrganizationId = organizerOrganizationProfile?.organizationId;
+
+  const bookerUrl = booking.eventType?.team
+    ? await getBookerBaseUrl(booking.eventType.team.parentId)
+    : await getBookerBaseUrl(organizerOrganizationId ?? null);
```
Comment (Udit-takkar) [inline packages/lib/payment/getBooking.ts:134]: same as above https://github.com/calcom/cal.com/pull/13893/files#r1525385023

Comment (Udit-takkar) [reply packages/lib/payment/getBooking.ts:134]: and other places

**Review by CarinaWolli — COMMENTED** [review]
I realized this can be done a lot easier and we can just use `reschedule/[uid]`, so we don't need any team slug or username. Ready for review now again 🙏

**Review by Udit-takkar — APPROVED** [review]
LGTM

**Pull request merged**